libs/vkd3d/swapchain.c

```diff
@@ -0,0 +1,2544 @@
+    return S_OK;
+}
+
+static BOOL d3d12_swapchain_is_present_mode_supported(struct d3d12_swapchain *swapchain,
+        VkPresentModeKHR present_mode)
+{
+    VkPhysicalDevice vk_physical_device = d3d12_swapchain_device(swapchain)->vk_physical_device;
+    const struct vkd3d_vk_device_procs *vk_procs = d3d12_swapchain_procs(swapchain);
+    VkPresentModeKHR *modes;
+    uint32_t count, i;
+    BOOL supported;
+    VkResult vr;
+
+    if (present_mode == VK_PRESENT_MODE_FIFO_KHR)
+        return TRUE;
+
+    if ((vr = vk_procs->vkGetPhysicalDeviceSurfacePresentModesKHR(vk_physical_device,
+            swapchain->vk_surface, &count, NULL)) < 0)
+    {
+        WARN("Failed to get count of available present modes, vr %d.\n", vr);
+        return FALSE;
+    }
+
+    supported = FALSE;
+
+    if (!(modes = vkd3d_calloc(count, sizeof(*modes))))
+        return FALSE;
+    if ((vr = vk_procs->vkGetPhysicalDeviceSurfacePresentModesKHR(vk_physical_device,
+            swapchain->vk_surface, &count, modes)) >= 0)
+    {
+        for (i = 0; i < count; ++i)
+        {
+            if (modes[i] == present_mode)
+            {
+                supported = TRUE;
+                break;
+            }
+        }
+    }
+    else
+    {
+        WARN("Failed to get available present modes, vr %d.\n", vr);
+    }
+    vkd3d_free(modes);
+
+    return supported;
+}
+
+static BOOL d3d12_swapchain_has_user_images(struct d3d12_swapchain *swapchain)
+{
+    return !!swapchain->vk_images[0];
+}
+
+static HRESULT d3d12_swapchain_create_user_buffers(struct d3d12_swapchain *swapchain, VkFormat vk_format)
+{
+    D3D12_HEAP_PROPERTIES heap_properties;
+    D3D12_RESOURCE_DESC resource_desc;
+    struct d3d12_resource* object;
+    HRESULT hr;
+    UINT i;
+
+    if (d3d12_swapchain_has_user_images(swapchain))
+        return S_OK;
+
+    heap_properties.Type = D3D12_HEAP_TYPE_DEFAULT;
+    heap_properties.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
+    heap_properties.MemoryPoolPreference = D3D12_MEMORY_POOL_UNKNOWN;
+    heap_properties.CreationNodeMask = 1;
+    heap_properties.VisibleNodeMask = 1;
+
+    resource_desc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+    resource_desc.Alignment = 0;
+    resource_desc.Width = swapchain->desc.Width;
+    resource_desc.Height = swapchain->desc.Height;
+    resource_desc.DepthOrArraySize = 1;
+    resource_desc.MipLevels = 1;
+    resource_desc.Format = dxgi_format_from_vk_format(vk_format);
+    resource_desc.SampleDesc.Count = 1;
+    resource_desc.SampleDesc.Quality = 0;
+    resource_desc.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN;
+    resource_desc.Flags = D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
+
+    for (i = 0; i < swapchain->desc.BufferCount; i++)
+    {
+        if (FAILED(hr = d3d12_committed_resource_create(d3d12_swapchain_device(swapchain),
+            &heap_properties, D3D12_HEAP_FLAG_NONE,
+            &resource_desc, D3D12_RESOURCE_STATE_COPY_SOURCE,
+            NULL, &object)))
```

Comment:
Formatting -- for multi-line function calls we have 8 spaces rather than 4, and this can probably be shortened to 3 lines as well rather than 4.

Reply:
Done did it